### PR TITLE
fix(scipy.linalg): toeplitz/hankel incorrectly propagate NaN via conv…

### LINKED ISSUE #38636: jax.scipy.linalg.toeplitz incorrectly propagates NaN to the entire output matrix
### Description

`jax.scipy.linalg.toeplitz` appears to incorrectly propagate `NaN` values when constructing a Toeplitz matrix from inputs containing nonfinite values.

A Toeplitz matrix should be constructed by selecting entries from the first column and first row. In this example, `row[0]` is `NaN`, but it should only affect positions that actually correspond to the overlapping diagonal value. It should not contaminate the whole output matrix.

However, JAX returns an all-`NaN` matrix, while the expected output contains finite values, `inf`, and `-0.0` in positions selected from `column` and `row`.

This looks like the implementation may be using arithmetic or masking in a way that allows `NaN` to propagate, instead of performing pure indexing/selection.

### Minimal reproducible example

```
#!/usr/bin/env python3
import os
import sys
import warnings

warnings.filterwarnings("ignore")
os.environ.setdefault("JAX_PLATFORMS", "cpu")

try:
    import numpy as np
    import jax.numpy as jnp
    import jax.scipy.linalg as jlinalg
except ImportError as e:
    print(f"missing dep: {e}")
    sys.exit(2)

column = np.array([np.nan, np.inf, -0.0], dtype=np.float32)
row = np.array([np.nan, 2.0, 3.0], dtype=np.float32)

expected = np.array(
    [
        [np.nan, 2.0, 3.0],
        [np.inf, np.nan, 2.0],
        [-0.0, np.inf, np.nan],
    ],
    dtype=np.float32,
)

out = np.asarray(jlinalg.toeplitz(jnp.asarray(column), jnp.asarray(row)))

print("jax:", out.tolist())
print("expected:", expected.tolist())

if not np.array_equal(out, expected, equal_nan=True):
    sys.exit(0)
sys.exit(1)
```

### Actual result

```
jax: [[nan, nan, nan], [nan, nan, nan], [nan, nan, nan]]
expected: [[nan, 2.0, 3.0], [inf, nan, 2.0], [-0.0, inf, nan]]
```

### Expected result

```
[[nan, 2.0, 3.0],
 [inf, nan, 2.0],
 [-0.0, inf, nan]]
```

For a Toeplitz matrix, the output should be formed by selecting values from `column` and `row`. `NaN` values should only appear in positions that correspond to an actual selected `NaN` input element. They should not propagate through unrelated finite, infinite, or signed-zero entries.

### System info (python version, jaxlib version, accelerator, etc.)

jax: 0.10.1
jaxlib: 0.10.1
numpy: 2.4.6
python: 3.13.13
platform: Ubuntu 24.04.3 LTS, Linux x86_64
accelerator: CPU backend

Note: GPU is present, but CUDA-enabled jaxlib is not installed, so JAX falls back to CPU.

### PR DESCRIPTION
**Pull request by Adityakk9031**
Fixes #38636.

jax.scipy.linalg.toeplitz returned an all-NaN matrix whenever any element of c or r was NaN, even when those values should only appear at specific diagonal positions.

python
# Before
jax.scipy.linalg.toeplitz([nan, inf, -0.], [nan, 2., 3.])
# → [[nan, nan, nan], [nan, nan, nan], [nan, nan, nan]]  
# After
# → [[nan, 2.0, 3.0], [inf, nan, 2.0], [-0., inf, nan]]  
Root Cause

_toeplitz (and _hankel) used lax.conv_general_dilated_patches — a convolution — to extract sliding windows from the concatenated input vector. Convolution involves arithmetic sums: NaN + x = NaN per IEEE 754, so any NaN in the kernel window poisoned every output position that window overlapped, turning the entire matrix NaN.

Fix

Toeplitz output cell A[i, j] maps to elems[ncols-1-i+j] for elems = c[::-1] ++ r[1:] in both triangles — a pure integer-indexed gather, not arithmetic. NaN isolation is guaranteed by construction.

python
# _toeplitz — 3 lines replacing 6
elems = jnp.concatenate((c[::-1], r[1:]))
i, j = jnp.arange(ncols)[:, None], jnp.arange(nrows)[None, :]
return elems[ncols - 1 - i + j]
# _hankel
v = jnp.concatenate((c, r[1:]))
return v[i + j]
The fix is also shorter and simpler (no reshape, no flip, no convolution), and produces identical results for all finite inputs.

Changes

File	Change
jax/_src/scipy/linalg.py	Replace conv_general_dilated_patches with index gather in _toeplitz and _hankel
tests/linalg_test.py	Add testToeplitzNaNPropagation + testHankelNaNPropagation regression tests
Tests — testToeplitzNaNPropagation reproduces the exact MRE, asserts exactly 3 NaNs on the main diagonal and zero elsewhere. testHankelNaNPropagation covers the same root cause in _hankel, asserting exactly 1 NaN at H[0,0].

References — Closes #38636

### CODE REVIEW

**Comment by Adityakk9031** [comment]
@bchetioui have look a please

**Comment by jakevdp** [comment]
This falls under #38651; please note that it may take a while for this to receive attention from the team.
